google/resource_cloudiot_registry.go

```diff
@@ -0,0 +1,411 @@
+							Required:         true,
+							DiffSuppressFunc: compareSelfLinkOrResourceName,
+						},
+					},
+				},
+			},
+			"state_notification_config": &schema.Schema{
+				Type:     schema.TypeMap,
+				Optional: true,
+				ForceNew: false,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"pubsub_topic_name": &schema.Schema{
+							Type:             schema.TypeString,
+							Required:         true,
+							DiffSuppressFunc: compareSelfLinkOrResourceName,
+						},
+					},
+				},
+			},
+			"mqtt_config": &schema.Schema{
+				Type:     schema.TypeMap,
+				Optional: true,
+				ForceNew: false,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"mqtt_enabled_state": &schema.Schema{
+							Type:     schema.TypeString,
+							Required: true,
+							ValidateFunc: validation.StringInSlice(
+								[]string{mqttStateUnspecified, mqttEnabled, mqttDisabled}, false),
+						},
+					},
+				},
+			},
+			"http_config": &schema.Schema{
+				Type:     schema.TypeMap,
+				Optional: true,
+				ForceNew: false,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"http_enabled_state": &schema.Schema{
+							Type:     schema.TypeString,
+							Required: true,
+							ValidateFunc: validation.StringInSlice(
+								[]string{httpStateUnspecified, httpEnabled, httpDisabled}, false),
+						},
+					},
+				},
+			},
+			"credentials": &schema.Schema{
+				Type:     schema.TypeList,
+				Optional: true,
+				ForceNew: false,
+				MaxItems: 10,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"public_key_certificate": &schema.Schema{
+							Type:     schema.TypeMap,
+							Optional: true,
+							Elem: &schema.Resource{
+								Schema: map[string]*schema.Schema{
+									"format": &schema.Schema{
+										Type:     schema.TypeString,
+										Required: true,
+										ValidateFunc: validation.StringInSlice(
+											[]string{unspecifiedCertFormat, x509CertificatePEM}, false),
+									},
+									"certificate": &schema.Schema{
+										Type:     schema.TypeString,
+										Required: true,
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func expandEventNotificationConfigs(configs []interface{}) []*cloudiot.EventNotificationConfig {
+	eventConfigs := make([]*cloudiot.EventNotificationConfig, len(configs))
+	for i, raw := range configs {
+		data := raw.(map[string]interface{})
+
+		eventConfigs[i] = &cloudiot.EventNotificationConfig{
+			PubsubTopicName: data["pubsub_topic_name"].(string),
+		}
+	}
+	return eventConfigs
+}
+
+func expandStateNotificationConfig(config map[string]interface{}) *cloudiot.StateNotificationConfig {
+	if v, ok := config["pubsub_topic_name"]; ok {
+		return &cloudiot.StateNotificationConfig{
+			PubsubTopicName: v.(string),
+		}
+	}
+	return nil
+}
+
+func expandMqttConfig(config map[string]interface{}) *cloudiot.MqttConfig {
```

Comment:
I'm not sure these are "expand"s, which we normally use to mean translating config lists -> go API object slices.  Do you think the naming convention should stretch to cover this case or do you have another idea for naming these?

Reply:
I just renamed the function to `buildMqttConfig` (and the others). Hoping that this fits better.